WalletWasabi.Tests/UnitTests/WabiSabi/Crypto/OwnershipProofTests.cs

```diff
@@ -0,0 +1,44 @@
+using NBitcoin;
+using WalletWasabi.Crypto;
+using WalletWasabi.Tests.Helpers;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.WabiSabi.Crypto
+{
+	/// <summary>
+	/// Tests for <see cref="OwnershipProof"/> class.
+	/// </summary>
+	public class OwnershipProofTests
+	{
+		[Fact]
+		public void EqualityTest()
+		{
+			using Key key = new();
+			uint256 roundHash = BitcoinFactory.CreateUint256();
+
+			// Request #1.
+			OwnershipProof request1 = CreateOwnershipProof(key, roundHash);
+
+			//// Request #2.
+			OwnershipProof request2 = CreateOwnershipProof(key, roundHash);
+
+			Assert.Equal(request1, request2);
+
+			//// Request #3.
+			using Key key2 = new();
+			OwnershipProof request3 = CreateOwnershipProof(key2, roundHash); // Key intentionally changed.
+			Assert.NotEqual(request1, request3);
+
+			//// Request #4.
+			OwnershipProof request4 = CreateOwnershipProof(key, roundHash: BitcoinFactory.CreateUint256()); // Round hash intentionally changed.
+			Assert.NotEqual(request1, request4);
+		}
+
+		/// <remarks>Each instance represents the same proof but a new object instance.</remarks>
+		public static OwnershipProof CreateOwnershipProof(Key key, uint256 roundHash)
+			=> OwnershipProof.GenerateCoinJoinInputProof(
+				key,
+				new OwnershipIdentifier(Key.Parse("5KbdaBwc9Eit2LrmDp1WfZd815StNstwHanbRrPpGGN6wWJKyHe", Network.Main), key.PubKey.WitHash.ScriptPubKey),
```

Comment:
if you mean 

```cs
new OwnershipIdentifier(key, key.PubKey.WitHash.ScriptPubKey),
```

Yes, I can change it. For the purposes of the `OwnershipProofTests` class, it should not matter.

Unfortunately, I don't really know what `identificationKey` in `OwnershipIdentifier` really is. Could you explain in a few sentences?

Reply:
Yep, that's what I meant!

The gist of it is since hardware wallets have constrained memory/state storage and cannot trust the host with state information, to avoid a number of attacks in the coinjoin setting when signing the transaction they need certainty about which inputs belong to them.

An ownership proof proves to the world that whoever made it is able to spend the coin (that's the BIP 322 part), but the SLIP 19 format for ownership proofs additionally allows hardware wallets to recognize their own inputs.

This is accomplished by deriving another key from the master key (SLIP 21 derivation), and using that as the secret in an HMAC. The HMAC output is indistinguishable from random bits unless the secret key is known, so it is unlinkable, but by  ensuring that ownership proofs are only generated with this identifier always using the same HMAC key, a hardware wallet can always recognize reliably which outputs belong to it.

This is required because without such a static key being used to identify things, there's an intractably large BIP 32 hierarchy to search through, recognizing that a `ScriptPubKey` on its own (without the derivation path) is too difficult for constrained devices.